### ISSUE TITLE
Post-Processing G-TAD For Custom Dataset :

### ISSUE THREAD

**Issue by lakshaymehra** [frostinassiky/gtad#10]
Hi,

I have successfully run the gtad_inferency.py script for my own dataset but would now like to post-process my results. I have the following doubts :

1. The gtad_postprocess.py script uses a numpy array called uNet_test.npy It is an array of shape (1574,101) which means the shape format is (number_of_test_videos, number_of_classes). It seems like it contains the classification results for each video in the test set. However, can you tell me how I can generate a similar file for my custom data set? Do I need to train a separate Classifier?

2. The gtad_postprocess.py script also generates GTAD proposals for each video. I am planning to use those proposals (instead of BSN proposals) to train PGCN to get the SOTA result as mentioned in your paper. Is that possible given that the format of GTAD proposals is different than BSN?

Thanks!

**Comment by frostinassiky** [comment]
1. Your understanding is correct. `uNet_test.npy` contains the classification results for each video in the test set. You can use any video classification model to produce the prediction. I recently used TSM (https://github.com/mit-han-lab/temporal-shift-module) to obtain the classification cores. If your dataset has different classes from THUMOS, You should train a separate Classifier.

**Comment by frostinassiky** [comment]
2. In my G-TAD+PGCN method, I modified the dataset class of PGCN, instead of saving G-TAD proposal as PGCN format. More concretely, you can update this function (https://github.com/Alvin-Zeng/PGCN/blob/46428daf719f7d655f19b3f27ed879bd88d91158/pgcn_dataset.py#L237) to load G-TAD result.

**Comment by lakshaymehra** [comment]
@Frostinassiky Thanks so much for your detailed response. I must admit that I am a fan of G-TAD and it's currently one of my favourite repositories on GitHub not only because the paper is the State of the Art For Temporal Action Localisation but also because you and @coolbay are so prompt and kind in responding to my doubts! Thanks a lot once again!

I had a bit of confusion with regards to a statement in the paper which says that:

In the Problem Formulation section,

"Following recent video action proposal generation
methods [5, 13, 15, 30], we construct our G-TAD model us-
ing feature sequences extracted from raw video frames. We
average the features of every σ consecutive frames and re-
fer to each set of the σ frames as a snippet."

In the implementation section, 

"For
THUMOS-14, the video features are extracted using TSN
model [44] pre-trained on Kinetics [57] with σ = 5."

In the repository, https://github.com/Frostinassiky/gtad/blob/7d88c497c3fbf83278012bdc29549d066f279fd3/gtad_lib/dataset.py#L208 
we are using the skip_video_frames = 5 as default and are basically down-sampling the features shape from (number_of_frames,1024) to (number_of_frames/5 ,1024). 

1. However, from what I understood in the above mentioned script, there is no implementation of "averaging the features of σ = 5 consecutive frames". Am I missing something or are we just picking every 5th frame's features and assuming that it is an estimate of the average?
2.  Also, I'm using I3D per-frame features as I found in the I3D paper that I3D pretrained on ImageNet and Kinetics is better than TSN and I found it's implementation to be much simpler than TSN. Does that affect G-TAD performance in any way as long as we have per-frame features as input to GTAD of shape (number_of_frames,1024)?

**Comment by coolbay** [comment]
Hi, @lakshaymehra ! 
1. You are right. According to the current implementation on Thumos, we directly take 1 frame for every 5 frames instead of averaging them. You can also try averaging them to see whether you can get better/worse performance. 
2. We also find that for a lot of methods, I3D features work better than TSN features. We didn't test on I3D features, but we guess that G-TAD should have even better performance using I3D features. If you can test G-TAD on I3D features, we would be very glad to hear from you about the performance.

**Comment by lakshaymehra** [comment]
Thanks a lot @coolbay !

**Comment by ZhengLeon** [comment]
Hi, @frostinassiky , @lakshaymehra 
The test set has only 213 videos, why there are 1574 items in uNet_test.npy file?
Thanks a lot!

**Comment by lakshaymehra** [comment]
Hi @ZhengLeon ,

The THUMOS'14 Data Set Contains 1574 Videos In The Test Set. You can find out more here : https://www.crcv.ucf.edu/THUMOS14/download.html

Hope that helps!

**Comment by ZhengLeon** [comment]
Thanks for you kind reply! @lakshaymehra

**Comment by destousok** [comment]
> 1. Your understanding is correct. `uNet_test.npy` contains the classification results for each video in the test set. You can use any video classification model to produce the prediction. I recently used TSM (https://github.com/mit-han-lab/temporal-shift-module) to obtain the classification cores. If your dataset has different classes from THUMOS, You should train a separate Classifier.

Hello  @frostinassiky ! I think that I cant understand completely the use of `uNet_test.npy' and of training a new different classifier for it. Can you give me a simple explanation for it? Thank you!

**Comment by JosephineRabbit** [comment]
> 1. In my G-TAD+PGCN method, I modified the dataset class of PGCN, instead of saving G-TAD proposal as PGCN format. More concretely, you can update this function (https://github.com/Alvin-Zeng/PGCN/blob/46428daf719f7d655f19b3f27ed879bd88d91158/pgcn_dataset.py#L237) to load G-TAD result.

Hi, @frostinassiky , could you kindly provide the mentioned function to load G-TAD result?

**Comment by frostinassiky** [comment]
Hi Josephine, I don't have this function any more.
